src/main/java/minijava/ir/optimize/Optimizer.java

```diff
@@ -52,23 +52,26 @@ static void optimize() {
             .dependsOn(controlFlowOptimizer, floatInTransformation)
             .build();
 
+    ProgramMetrics metrics = ProgramMetrics.analyse(firm.Program.getGraphs());
+    Inliner inliner = new Inliner(metrics, true);
     while (true) {
       for (Graph graph : firm.Program.getGraphs()) {
         perGraphFramework.optimizeUntilFixedpoint(graph);
       }
 
       // Here comes the interprocedural stuff... This is method is really turning into a mess
       boolean hasChanged = false;
-      ProgramMetrics metrics = ProgramMetrics.analyse(firm.Program.getGraphs());
-      Optimizer inliner = new Inliner(metrics);
       for (Graph graph : firm.Program.getGraphs()) {
         hasChanged |= inliner.optimize(graph);
         unreachableCodeRemover.optimize(graph);
-        metrics.updateGraphInfo(graph);
         Cli.dumpGraphIfNeeded(graph, "after-inlining");
       }
       if (!hasChanged) {
-        break;
+        if (inliner.onlyLeafs) {
+          inliner = new Inliner(metrics, false);
```

Comment:
Now I'm a little confused. Didn't you say you wanted a flag?
This is really close to simulating object orientation with enums. Well, the flag is basically doing the same thing, but if you _do_ want extensibility we should, in my opinion, go with real object orientation instead of enums.